hail/hail/src/is/hail/expr/ir/IR.scala

```diff
@@ -66,65 +61,12 @@ sealed trait IR extends BaseIR {
 
 package defs {
 
-  import is.hail.expr.ir.defs.ArrayZipBehavior.ArrayZipBehavior
-
-  sealed trait TypedIR[T <: Type] extends IR {
+  trait TypedIR[T <: Type] extends IR {
     override def typ: T = tcoerce[T](super.typ)
   }
 
   // Mark Refs and constants as IRs that are safe to duplicate
-  sealed trait TrivialIR extends IR
-
-  object Literal {
-    def coerce(t: Type, x: Any): IR = {
-      if (x == null)
-        return NA(t)
-      t match {
-        case TInt32 => I32(x.asInstanceOf[Number].intValue())
-        case TInt64 => I64(x.asInstanceOf[Number].longValue())
-        case TFloat32 => F32(x.asInstanceOf[Number].floatValue())
-        case TFloat64 => F64(x.asInstanceOf[Number].doubleValue())
-        case TBoolean => if (x.asInstanceOf[Boolean]) True() else False()
-        case TString => Str(x.asInstanceOf[String])
-        case _ => Literal(t, x)
-      }
-    }
-  }
-
-  final case class Literal(_typ: Type, value: Annotation) extends IR {
-    require(!CanEmit(_typ))
-    require(value != null)
-    // expensive, for debugging
-    //   require(SafeRow.isSafe(value))
-    //   assert(_typ.typeCheck(value), s"literal invalid:\n  ${_typ}\n  $value")
-  }
-
-  object EncodedLiteral {
-    def apply(codec: AbstractTypedCodecSpec, value: Array[Array[Byte]]): EncodedLiteral =
-      EncodedLiteral(codec, new WrappedByteArrays(value))
-
-    def fromPTypeAndAddress(pt: PType, addr: Long, ctx: ExecuteContext): IR = {
-      pt match {
-        case _: PInt32 => I32(Region.loadInt(addr))
-        case _: PInt64 => I64(Region.loadLong(addr))
-        case _: PFloat32 => F32(Region.loadFloat(addr))
-        case _: PFloat64 => F64(Region.loadDouble(addr))
-        case _: PBoolean => if (Region.loadBoolean(addr)) True() else False()
-        case ts: PString => Str(ts.loadString(addr))
-        case _ =>
-          val etype = EType.defaultFromPType(ctx, pt)
-          val codec = TypedCodecSpec(etype, pt.virtualType, BufferSpec.wireSpec)
-          val bytes = codec.encodeArrays(ctx, pt, addr)
-          EncodedLiteral(codec, bytes)
-      }
-    }
-  }
-
-  final case class EncodedLiteral(codec: AbstractTypedCodecSpec, value: WrappedByteArrays)
-      extends IR {
-    require(!CanEmit(codec.encodedVirtualType))
-    require(value != null)
-  }
+  trait TrivialIR extends IR
```

Comment:
why remove `sealed`?

Reply:
`sealed` requires all subclasses be defined in the same file as the trait. Anyways, I think match statements that exhaustively list every IR node are a code smell, and I'm not sure we actually have any that benefit from exhaustivity checking.

Reply:
Oh I see, thanks.